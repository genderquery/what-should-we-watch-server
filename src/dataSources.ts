import { RequestOptions, RESTDataSource } from "apollo-datasource-rest";

export class TmdbDataSource extends RESTDataSource {
  private apiToken: string;

  constructor(apiToken: string) {
    super();
    this.apiToken = apiToken;
    this.baseURL = "https://api.themoviedb.org/";
  }

  willSendRequest(request: RequestOptions) {
    request.headers.set("Authorization", `Bearer ${this.apiToken}`);
  }

  static movieReducer(source: any) {
    return {
      id: source.id,
      title: source.title,
      releaseDate: source.release_date,
      overview: source.overview,
    };
  }

  async getMovieById(id: number) {
    const result = await this.get(`3/movie/${id}`);
    return TmdbDataSource.movieReducer(result);
  }
}

export default () => ({
  tmdb: new TmdbDataSource(process.env.TMDB_API_TOKEN!),
});
