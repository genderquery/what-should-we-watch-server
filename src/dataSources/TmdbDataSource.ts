import { RequestOptions, RESTDataSource } from "apollo-datasource-rest";

export class TmdbDataSource<Context> extends RESTDataSource<Context> {
  private apiToken: string;

  constructor(apiToken: string) {
    super();
    this.apiToken = apiToken;
    this.baseURL = "https://api.themoviedb.org/";
  }

  willSendRequest(request: RequestOptions): void {
    request.headers.set("Authorization", `Bearer ${this.apiToken}`);
  }

  static movieReducer(source: any) {
    return {
      id: source.id,
      title: source.title,
      releaseDate: source.release_date,
      overview: source.overview,
    };
  }

  async getMovieById(id: string) {
    const result = await this.get(`3/movie/${id}`);
    return TmdbDataSource.movieReducer(result);
  }
}

export default TmdbDataSource;
