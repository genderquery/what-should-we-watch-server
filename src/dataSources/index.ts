import { DataSource } from "apollo-datasource";
import { TmdbDataSource } from "./TmdbDataSource";

export interface DataSources<Context>
  extends Record<string, DataSource<Context>> {
  tmdb: TmdbDataSource<Context>;
}

const dataSources = () => ({
  tmdb: new TmdbDataSource(process.env.TMDB_API_TOKEN!),
});

export { TmdbDataSource };
export default dataSources;
