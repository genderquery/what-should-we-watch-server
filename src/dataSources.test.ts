import { RequestOptions } from "apollo-datasource-rest";
import { Headers, URLSearchParams } from "apollo-server-env";
import { TmdbDataSource } from "./dataSources";

const mockMovie = {
  adult: false,
  backdrop_path: "/fNG7i7RqMErkcqhohV2a6cV1Ehy.jpg",
  belongs_to_collection: {
    id: 2344,
    name: "The Matrix Collection",
    poster_path: "/lh4aGpd3U9rm9B8Oqr6CUgQLtZL.jpg",
    backdrop_path: "/bRm2DEgUiYciDw3myHuYFInD7la.jpg",
  },
  budget: 63000000,
  genres: [
    {
      id: 28,
      name: "Action",
    },
    {
      id: 878,
      name: "Science Fiction",
    },
  ],
  homepage: "http://www.warnerbros.com/matrix",
  id: 603,
  imdb_id: "tt0133093",
  original_language: "en",
  original_title: "The Matrix",
  overview:
    "Set in the 22nd century, The Matrix tells the story of a computer hacker who joins a group of underground insurgents fighting the vast and powerful computers who now rule the earth.",
  popularity: 51.055,
  poster_path: "/dXNAPwY7VrqMAo51EKhhCJfaGb5.jpg",
  production_companies: [
    {
      id: 79,
      logo_path: "/tpFpsqbleCzEE2p5EgvUq6ozfCA.png",
      name: "Village Roadshow Pictures",
      origin_country: "US",
    },
    {
      id: 372,
      logo_path: null,
      name: "Groucho II Film Partnership",
      origin_country: "",
    },
    {
      id: 1885,
      logo_path: "/xlvoOZr4s1PygosrwZyolIFe5xs.png",
      name: "Silver Pictures",
      origin_country: "US",
    },
    {
      id: 174,
      logo_path: "/IuAlhI9eVC9Z8UQWOIDdWRKSEJ.png",
      name: "Warner Bros. Pictures",
      origin_country: "US",
    },
  ],
  production_countries: [
    {
      iso_3166_1: "AU",
      name: "Australia",
    },
    {
      iso_3166_1: "US",
      name: "United States of America",
    },
  ],
  release_date: "1999-03-30",
  revenue: 463517383,
  runtime: 136,
  spoken_languages: [
    {
      iso_639_1: "en",
      name: "English",
    },
  ],
  status: "Released",
  tagline: "Welcome to the Real World.",
  title: "The Matrix",
  video: false,
  vote_average: 8.1,
  vote_count: 17671,
};

const get = jest.fn<any, any>();

jest.mock("apollo-datasource-rest", () => ({
  __esModule: true,
  RESTDataSource: class {
    get = get
  },
}));

test("movieReducer", () => {
  const result = TmdbDataSource.movieReducer(mockMovie);
  expect(result).toMatchSnapshot();
});


test("sends API token", async () => {
  const tmdb = new TmdbDataSource("mytoken");
  const request: RequestOptions = {
    path: "/",
    params: new URLSearchParams(),
    headers: new Headers(),
  };
  tmdb.willSendRequest(request);
  expect(request.headers.get("Authorization")).toBe("Bearer mytoken");
});

test("getMovieById", async () => {
  get.mockReset()
  get.mockReturnValueOnce(mockMovie)

  const tmdb = new TmdbDataSource("mytoken");
  const result = await tmdb.getMovieById(506);

  expect(get).toHaveBeenCalledWith('3/movie/506')
  expect(result).toMatchSnapshot()
});
