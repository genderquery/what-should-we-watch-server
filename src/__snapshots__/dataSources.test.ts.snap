// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMovieById 1`] = `
Object {
  "id": 603,
  "overview": "Set in the 22nd century, The Matrix tells the story of a computer hacker who joins a group of underground insurgents fighting the vast and powerful computers who now rule the earth.",
  "releaseDate": "1999-03-30",
  "title": "The Matrix",
}
`;

exports[`movieReducer 1`] = `
Object {
  "id": 603,
  "overview": "Set in the 22nd century, The Matrix tells the story of a computer hacker who joins a group of underground insurgents fighting the vast and powerful computers who now rule the earth.",
  "releaseDate": "1999-03-30",
  "title": "The Matrix",
}
`;
